sync/sync_tail.go

```diff
@@ -0,0 +1,216 @@
+package sync
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/celestiaorg/go-header"
+)
+
+// subjectiveTail returns the current actual Tail header.
+// Lazily fetching it if it doesn't exist locally or moving it to a different height.
+// Moving is done if either parameters are changed or tail moved outside a pruning window.
+func (s *Syncer[H]) subjectiveTail(ctx context.Context, head H) (H, error) {
+	oldTail, err := s.store.Tail(ctx)
+	if err != nil && !errors.Is(err, header.ErrEmptyStore) {
+		return oldTail, err
+	}
+
+	newTail, err := s.renewTail(ctx, oldTail, head)
+	if err != nil {
+		return oldTail, fmt.Errorf("updating tail: %w", err)
+	}
+
+	if err := s.moveTail(ctx, oldTail, newTail); err != nil {
+		return oldTail, fmt.Errorf(
+			"moving tail from %d to %d: %w",
+			oldTail.Height(),
+			newTail.Height(),
+			err,
+		)
+	}
+
+	return newTail, nil
+}
+
+// renewTail resolves the new actual tail header respecting Syncer parameters.
+func (s *Syncer[H]) renewTail(ctx context.Context, oldTail, head H) (newTail H, err error) {
+	switch tailHash := s.tailHash(oldTail); tailHash {
+	case nil:
```

Comment:
IMO still a bit confusing here but better than last iteration. I don't immediately know what would make it cleaner but it's good for now.

Reply:
Yeah, the best i've came up so far